FHIR/fhir-package-loader
# Add Browser support for FPL

**Description:** This PR adds support for using FPL in a browser and using the browser's IndexedDB as a PackageCache.

The `BrowserBasedPackageCache` class is a new PackageCache implementation that can be used to store resources in the browser's IndexedDB. Because the IndexedDB API is asynchronous, the `BrowserBasedPackageCache` does some extra work to ensure the synchronous methods of the PackageCache interface still work. There is an async `initialize` function that looks in the IndexedDB to check what packages are already loaded and saves that information to a local map, and the `cachePackageTarball` function stores resources in IndexedDB and to the local map. The local map is then used for the rest of the methods, like `isPackageInCache`, as those are all synchronous.

There were also a few other changes that were needed when using in a browser, some of which I'd like your opinion on if there's a better way around it:

- `https.get` vs `axios.get`: I could not figure out how to have `axios.get` return something that could be unzipped in the browser. I suspect I must have missed something, but I ended up using `https.get`, which is what FSH Online currently uses. If anyone reviewing this has an idea of how to configure axios to get it to work with the `tar-stream` library or anything else to successfully unzip and get the data without a filesystem, let me know.
- `useHttps` option in `FHIRRegistryClient`: When I needed to use `https.get` instead of `axios.get`, I had originally created a separate RegistryClient (a `BrowserFHIRRegistryClient`), but all the code was the same except for the actual GET request. Instead, I ended up adding an option to the existing FHIRRegistryClient in an effort to duplicate as little code as possible. I'm not sure if this is the best approach, so let me know what you think.
- `nock` for the `https.get` calls: I could not figure out how to use Jest and spyOn and mocked implementations to successfully return responses that were like what `https.get` returns. I ended up using the same approach that FSH Online used, which used `nock`. That does add another dependency, and it feels like Jest _should_ be able to do the same, so if you know of another way around using `nock`, let me know.
- `initialize(initSql)` in `SQLJSPackageDB`: It turns out that I needed to pass a configuration option into `initSqlJs()` when initializing in the browser, so I needed to be able to initialize it in FSH Online and pass it in to `SQLJSPackageDB`.

**Testing Instructions:**
Make sure the tests reflect the expected functionality. I also pushed up a [new-fpl](https://github.com/FHIR/FSHOnline/tree/new-fpl) branch on FSH Online that uses this branch. Feel free to try out using FSH Online and making sure you can download and load packages.

**Related Issue:**
Fixes #29
